base/sparse/sparsematrix.jl

```diff
@@ -2532,23 +2525,59 @@ setindex!(A::SparseMatrixCSC, x, ::Colon, ::Colon) = setindex!(A, x, 1:size(A, 1
 setindex!(A::SparseMatrixCSC, x, ::Colon, j::Union{Integer, AbstractVector}) = setindex!(A, x, 1:size(A, 1), j)
 setindex!(A::SparseMatrixCSC, x, i::Union{Integer, AbstractVector}, ::Colon) = setindex!(A, x, i, 1:size(A, 2))
 
-setindex!{Tv,T<:Integer}(A::SparseMatrixCSC{Tv}, x::Number, I::AbstractVector{T}, J::AbstractVector{T}) =
-    (0 == x) ? spdelete!(A, I, J) : spset!(A, convert(Tv,x), I, J)
-
-function spset!{Tv,Ti<:Integer}(A::SparseMatrixCSC{Tv}, x::Tv, I::AbstractVector{Ti}, J::AbstractVector{Ti})
-    !issorted(I) && (I = sort(I))
-    !issorted(J) && (J = sort(J))
-
-    m, n = size(A)
-    lenI = length(I)
-
-    if (!isempty(I) && (I[1] < 1 || I[end] > m)) || (!isempty(J) && (J[1] < 1 || J[end] > n))
+function setindex!{TvA,Tx<:Number,TiI<:Integer,TiJ<:Integer}(A::SparseMatrixCSC{TvA}, x::Tx,
+        I::AbstractVector{TiI}, J::AbstractVector{TiJ})
+    if isempty(I) || isempty(J); return A; end
+    if !issorted(I); I = sort(I); end
+    if !issorted(J); J = sort(J); end
+    if (I[1] < 1 || I[end] > A.m) ||
+         (J[1] < 1 || J[end] > A.n)
         throw(BoundsError(A, (I, J)))
     end
-
-    if isempty(I) || isempty(J)
-        return A
+    if x == 0
+        _spsetz_setindex!(A, I, J)
+    else
+        _spsetnz_setindex!(A, convert(TvA, x), I, J)
+    end
+end
+"""
+Helper method for immediately preceding setindex! method. For all (i,j) such that i in I and
+j in J, assigns zero to A[i,j] if A[i,j] is a presently-stored entry, and otherwise does nothing.
+"""
+function _spsetz_setindex!{TvA,TiI<:Integer,TiJ<:Integer}(A::SparseMatrixCSC{TvA},
+        I::AbstractVector{TiI}, J::AbstractVector{TiJ})
+    lengthI = length(I)
+    for j in J
+        coljAfirstk = A.colptr[j]
+        coljAlastk = A.colptr[j+1] - 1
+        kA = coljAfirstk
+        kI = 1
+        while kA <= coljAlastk && kI <= lengthI
+            # On the one hand, this while structure performs more loads than it should.
```

Comment:
Rewritten avoiding the excess loads. Slightly more complex but still quite readable it seems. Thoughts? Thanks!